substrate-archive/src/actors.rs

```diff
@@ -92,44 +91,46 @@ where
     }
 }
 
-struct Actors<B: BlockT + Unpin>
+struct Actors<B: BlockT + Unpin, D: ReadOnlyDB + 'static>
 where
     B::Hash: Unpin,
     NumberFor<B>: Into<u32>,
 {
     storage: Address<workers::StorageAggregator<B>>,
-    blocks: Address<workers::BlocksIndexer<B>>,
+    blocks: Address<workers::BlocksIndexer<B, D>>,
     metadata: Address<workers::Metadata<B>>,
     db_pool: Address<ActorPool<DatabaseActor<B>>>,
 }
 
 /// Control the execution of the indexing engine.
 /// Will exit on Drop.
-pub struct System<B, R, C>
+pub struct System<B, R, C, D>
 where
+    D: ReadOnlyDB + 'static,
```

Comment:
you can remove all the trait bounds on the struct and just add them on each impl.